pkg/spoke/controllers/deletioncontroller/deletion_controller.go

```diff
@@ -112,41 +134,95 @@ func (m *StaleManifestDeletionController) syncManifestWork(ctx context.Context,
 	_, err := m.manifestWorkClient.UpdateStatus(ctx, manifestWork, metav1.UpdateOptions{})
 	return err
 }
 
-// deleteAppliedResources deletes all applied resources in the given resource meta array.
-func (m *StaleManifestDeletionController) deleteAppliedResources(resources []workapiv1.AppliedManifestResourceMeta) []error {
+// deleteAppliedResources deletes all given applied resources and returns those
+// which are deleted but not finalized yet
+func (m *StaleManifestDeletionController) deleteAppliedResources(resources []workapiv1.AppliedManifestResourceMeta) ([]workapiv1.AppliedManifestResourceMeta, []error) {
+	var deletingResources []workapiv1.AppliedManifestResourceMeta
 	var errs []error
+
 	for _, resource := range resources {
 		gvr := schema.GroupVersionResource{Group: resource.Group, Version: resource.Version, Resource: resource.Resource}
-		err := m.spokeDynamicClient.Resource(gvr).Namespace(resource.Namespace).Delete(context.TODO(), resource.Name, metav1.DeleteOptions{})
-		switch {
-		case errors.IsNotFound(err):
-			// no-oop
-		case err != nil:
-			errs = append(errs, fmt.Errorf("Failed to delete resource %v with key %s/%s: %w", gvr, resource.Namespace, resource.Name, err))
+		u, err := m.spokeDynamicClient.
+			Resource(gvr).
+			Namespace(resource.Namespace).
+			Get(context.TODO(), resource.Name, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			klog.V(2).Infof("Resource %v with key %s/%s is removed Successfully", gvr, resource.Namespace, resource.Name)
 			continue
 		}
-		klog.V(2).Infof("Successfully delete resource %v with key %s/%s", gvr, resource.Namespace, resource.Name)
+
+		if err != nil {
+			errs = append(errs, fmt.Errorf(
+				"Failed to get resource %v with key %s/%s: %w",
+				gvr, resource.Namespace, resource.Name, err))
+			deletingResources = append(deletingResources, resource)
+			continue
+		}
+
+		// leave deleting resource untouched
+		if deletionTimestamp := u.GetDeletionTimestamp(); deletionTimestamp != nil && !deletionTimestamp.IsZero() {
+			// set the UID even if the resource is deleted by other client
+			if resource.UID == "" {
+				resource.UID = string(u.GetUID())
+			}
+			deletingResources = append(deletingResources, resource)
+			continue
+		}
+
+		if resource.UID != "" {
+			// Since the case meets the conditions belows should never happen,
+			//   a) resource exists
+			//   b) resource is not deleting
+			//   c) UID of applied resource is set
+			//   d) UID of applied resource equals resource UID
+			// the resource UID must be different with UID of applied resource, that means this resource must
+			// have already been deleted and then re-created by other client. Manifest work should not track it
+			// any longer
+			continue
+		}
+
+		// delete the resource which exists and is not being deleted
+		err = m.spokeDynamicClient.
+			Resource(gvr).
+			Namespace(resource.Namespace).
+			Delete(context.TODO(), resource.Name, metav1.DeleteOptions{})
```

Comment:
add a UID precondition here and check the appropriate error if the precondition fails.